clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -643,10 +644,19 @@ private KafkaConsumer(ConsumerConfig config,
             List<ConsumerInterceptor<K, V>> interceptorList = (List) (new ConsumerConfig(userProvidedConfigs)).getConfiguredInstances(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG,
                     ConsumerInterceptor.class);
             this.interceptors = interceptorList.isEmpty() ? null : new ConsumerInterceptors<>(interceptorList);
+
+            // the maximum time we can block in a call to NetworkClient.poll is limited by the frequency
+            // of periodic events (heartbeats and autocommits)
+            boolean enableAutoCommit = config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+            int autoCommitIntervalMs = config.getInt(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG);
+            int heartbeatIntervalMs = config.getInt(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG);
+            this.maxPollTimeoutMs = (enableAutoCommit ? Math.min(autoCommitIntervalMs, heartbeatIntervalMs) : heartbeatIntervalMs) / 2;
```

Comment:
I think `maxPollTimeout` was a bad idea. Whatever you set it to, it can end up delaying a needed heartbeat. The timeout really needs to be adjusted dynamically according to when the next heartbeat/autocommit is needed.